numalogic/registry/redis_registry.py

```diff
@@ -0,0 +1,247 @@
+import logging
+import time
+from typing import Optional, Sequence, Dict, Any
+
+from redis.client import Redis
+from redis.exceptions import RedisError
+
+from numalogic.registry import ArtifactManager, ArtifactData
+from numalogic.registry.serialize import loads, dumps
+from numalogic.tools.exceptions import ModelKeyNotFound
+from numalogic.tools.types import Artifact
+
+_LOGGER = logging.getLogger()
+
+
+class RedisRegistry(ArtifactManager):
+    """
+    Model saving and loading using Redis Registry.
+    Args:
+        client: Take in the reids client already established/created
+        ttl: Total Time to Live for the key when saving in redis (dafault = 5000)
+
+    Examples
+    --------
+    >>> import redis
+    >>> from numalogic.models.autoencoder.variants import VanillaAE
+    >>> from numalogic.registry.redis_registry import RedisRegistry
+    >>> r = redis.StrictRedis(host='127.0.0.1', port=6379)
+    >>> cli = r.client()
+    >>> registry = RedisRegistry(client=cli)
+    >>> skeys = ['c', 'a']
+    >>> dkeys = ['d', 'a']
+    >>> model = VanillaAE(10)
+    >>> registry.save(skeys, dkeys, artifact=model, **{'lr': 0.01})
+    >>> registry.load(skeys, dkeys, artifact=model)
+    """
+
+    __slots__ = ("client", "ttl")
+    _HOST = None
+
+    def __new__(
+        cls,
+        client: Redis = None,
+        ttl: int = 50000,
+        *args,
+        **kwargs,
+    ):
+        instance = super().__new__(cls, *args, **kwargs)
+        if not cls._HOST:
+            cls._HOST = "None"
+        return instance
+
+    def __init__(
+        self,
+        client: Redis = None,
+        ttl: int = 50000,
+    ):
+        if client:
+            super().__init__("None")
+            self.client = client
+        else:
+            raise ValueError("Missing Redis Client")
+        self.ttl = ttl
+
+    @staticmethod
+    def construct_key(*keys: [Sequence[str]]) -> str:
+        """
+        Returns a single key comprising static and dynamic key fields.
+        Args:
+            keys: Sequence of strings sequence combined to form key
+        Returns:
+            key
+        """
+        key = "::".join([":".join(key) for key in keys])
+        return key
+
+    @staticmethod
+    def get_version(key: str) -> str:
+        """
+        get version number from the string
+        Args:
+            key: key
+
+        Returns:
+            version
+        """
+        version_number = key.split("::")
```

Comment:
what if the key doesn't have `::` chars?

Reply:
It will definitely have "::".
The redis registry code introduces it while saving a model